Add use-page config option

jannschu opened the pull request:
The option allows to use a different page for the section page instead
of the first child.

oprypin commented on README.md:
```diff
@@ -61,6 +61,21 @@ To make writing this kind of `nav` more natural ([in YAML there's no better opti
+The value is the name of the page to use, in this case `index.md`. If a child
+with that name does not exist no section page is generated.
+
 [literate-nav]: https://oprypin.github.io/mkdocs-literate-nav/
```
This link reference should stay connected to its section above

oprypin commented on mkdocs_section_index/plugin.py:
```diff
@@ -16,6 +16,10 @@
 
 class SectionIndexPlugin(BasePlugin):
+    config_scheme = (
+        ('page', mkdocs.config.config_options.Type(str, default=None)),
```
So is it 'page' or 'use-page'?

oprypin replied:
The code needs to be reformatted.

`black -l100 .` would do it.

oprypin commented on README.md:
````diff
@@ -61,6 +61,21 @@ To make writing this kind of `nav` more natural ([in YAML there's no better opti
+```yaml
+plugins:
+  - section-index:
+    use-page: index
````
* Underscores are much more typical for options.
* For some reason I think it's really about files, not pages.
* Using just the stem of the file name is confusing.
* Needs an extra indent.

With all that together, this is what I would do:

```suggestion
      index_page: index.md
```

oprypin replied:
Oh sorry, I wrote this wrong.
I would really prefer to name it <code>index_<b>file</b></code>
```suggestion
      index_file: index.md
```

oprypin commented on mkdocs_section_index/plugin.py:
```diff
@@ -24,7 +28,16 @@ def on_nav(self, nav: Navigation, config, files) -> Navigation:
+                    page = section.children[0]
+                else:
+                    for child in section.children:
+                        if child.file.name == name:
```
Tip: to implement this as file-based, I believe you'd use `os.path.basename(child.file.src_path)`

oprypin commented on pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "mkdocs-section-index"
-version = "0.3.0"
+version = "0.4.0"
```
My projects are maintained in a way that does not accept a version change within a pull request.

oprypin reviewed (changes requested):
Thanks for the contribution. Overall it is good. But I have sooo many nitpicks.


In addition to the rest, this change should have tests along with it.
And even the existing tests are failing now.

To go along with this line
https://github.com/oprypin/mkdocs-section-index/blob/6e7b33e58312764ddb28f1aeaff64e1cbe75f3d7/tests/test_plugin.py#L74 you'd probably need
```python
plugin.config = {"index_file": golden.get("index_file")}
```

You will need to run tests. For that you can either run `poetry install; .tools/hooks/pre-commit`, or install dependencies directly and run `pytest` however you want.

jannschu commented:
Just saw you comment after my last push, I'm on it :)

jannschu commented:
I addressed your comments, but I have the feeling that `previous_page` is now wrong if `index_page` does not pick the first child. I'll add a test case to be sure.

oprypin commented:
Oh wow that's a lot of logic now :(
Do you think anyone will actually need the ability to pick index pages from the middle?
The option could just require it to be `index.md`. Doesn't that completely cover your use case?

jannschu commented:
> Oh wow that's a lot of logic now :(
> Do you think anyone will actually need the ability to pick index pages from the middle?
> The option could just require it to be `index.md`. Doesn't that completely cover your use case?

Actually `index.md` would be more than enough for me. So I'm fine with restricting it to that, but the logic is basically "fix a doubly linked list".

I'd even argue that if there is no index.md the first page should not become the section page by default since it then depends on alphabetical order but there might be use cases for that.

oprypin commented on tests/test_plugin.py:
```diff
@@ -15,39 +15,81 @@
+        if nav_src == "explicit_nav"
+        else None,
     )
+    if nav_src == "derived_nav" and index_file != "default_index_file":
```
Instead of adding actual files to the real example and branching all of this logic, I think the test should be separate and new.

If you need to use files from the real example, you can first create a directory for it then copy those files and add more.

```python
shutil.copytree(example_dir / "docs", str(tmpdir))

noindex_dir = (tmpdir / "docs").mkdir("z_noindex")
(noindex_dir / "a.md").write_text("# a")
```

then change `docs_dir=tmp_dir,`


Or if you don't need the real example, just don't copy the files

jannschu replied:
Sounds good

jannschu commented:
Just thougth about the case where a previous plugin might have reordered items such that `index.md` is no longer first. This could lead to unexpected results if we just check if the first child is `index.md` with that option. But we could omit a warning in such a case, but actually moving these element in a doubly linked list seems rather clean to me. But it's on you of course, just my final thoughts on this.

By the way if a plugin reorders children _after_ this plugin results are most likely wrong. We could also omit a warning maybe if popular plugins which could do that are behind this plugin.

oprypin commented:
The warning is there

https://github.com/oprypin/mkdocs-section-index/blob/7a8b4ac7c87716c2d6a10b672de21d1f5c299bb0/mkdocs_section_index/plugin.py#L53-L56

Though it doesn't detect subtle reorderings. But I think it's not unreasonable to expect the plugins to coexist anyway. Currently all mav reordering plugins I know  rewrite the whole nav.

solonovamax commented:
What's the status on this? Because it fixes #6, which is a very annoying issue for me.

oprypin commented:
Hmm the status is that I'm a bad human. I didn't follow up on this.
The feature is needed, but I wasn't sure how exactly it should look like.
